docs/plugins/plugin-api/asyncio.rst

```diff
@@ -0,0 +1,148 @@
+---------------------------
+
+Motivation
+##########
+
+A content unit that requires multiple files to be all downloaded before the content unit can be
+saved is a common problem. Consider a content unit foo, that requires three files, A, B, and C. One
+option is to use the :ref:`DownloaderFactory <download-factory>` to generate a downloader for each URL
+(A, B, C) and wait for those downloads to complete and then save the content unit foo and its
+associated Artifacts. The issue with this approach is that you also want to download other content
+units, e.g. a unit named bar with files (D, E, and F) and while waiting on A, B, and C you are not
+also downloading D, E, and F in parallel.
+
+GroupDownloader Overview
+########################
+
+The GroupDownloader allows you to schedule a :class:`~pulpcore.plugin.download.asyncio.Group`
+of downloads in a way that results are returned when the entire Group is ready instead of
+download-by-download. This is significant because multiple downloads from multiple groups still run
+in parallel. See the examples below.
+
+.. autoclass:: pulpcore.plugin.download.asyncio.GroupDownloader
+    :members:
+
+.. autoclass:: pulpcore.plugin.download.asyncio.Group
+    :members:
+
+.. _exception-handling:
+
+Exception Handling
+------------------
+
+All downloaders are expected to handle recoverable errors automatically. When an unrecoverable error
+occurs it can be one of two types of errors, i.e. one of the
+:ref:`validation exceptions <validation-exceptions>`, or a protocol specific error. A validation
+error example would be if a download size is expected to be 1945 bytes but it is actually 1990
+bytes, a :class:`~pulpcore.plugin.download.asyncio.SizeValidationError` would be raised. An example
+of a protocol specific error would be an HTTP 403 response.
+
+In both cases these are fatal exceptions and should likely be recorded with the
+:meth:`~pulpcore.plugin.tasking.Task.append_non_fatal_error` interface. A fatal exception on a single
+download likely does not cause an entire sync to fail, so a downloader's fatal exception is recorded
+as a non-fatal exception on the task. Plugin writers can also opt to halt the entire task by
+allowing the exception be uncaught and propogate up.
+
+.. _http-downloader:
+
+HttpDownloader
+--------------
+
+This downloader is an asyncio-aware parallel downloader which is the default downloader produced by
+the :ref:`download-factory` and the :ref:`group-downloader` For urls starting with `http://` or
+`https://`.
+
+.. autoclass:: pulpcore.plugin.download.asyncio.HttpDownloader
+    :members:
+
+.. _file-downloader:
+
+FileDownloader
+--------------
+
+This downloader is an asyncio-aware parallel file reader which is the default downloader produced by
+the :ref:`download-factory` and the :ref:`group-downloader` for urls starting with file://
+
+.. autoclass:: pulpcore.plugin.download.asyncio.FileDownloader
+    :members:
+
+.. _base-downloader:
+
+BaseDownloader
```

Comment:
IIRC, in your review of the _futures_ download PR, you explicitly requested that the `Download` base class be removed from the documentation because plugin writers were not expected to create customer downloaders or override behavior.  Do you still feel that way?  If so should `BaseDownloader` be removed?  Or, should I submit a PR to include the _futures_ base class as well?

Reply:
Yes, thanks for bringing this up. I think opening a PR to include the futures base class would be ideal. I think the use cases that require subclassing are in the must-have list of requirements and since they are both ready we should include them sooner (like now). Sound ok?

Reply:
Sounds good.